paddle/fluid/operators/assert_op.cc

```diff
@@ -0,0 +1,110 @@
+/* Copyright (c) 2020 PaddlePaddle Authors. All Rights Reserved.
+
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+   http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License. */
+
+#include "paddle/fluid/framework/op_registry.h"
+#include "paddle/fluid/framework/var_type.h"
+#include "paddle/fluid/operators/controlflow/while_op_helper.h"
+#include "paddle/fluid/operators/tensor_formatter.h"
+
+const char kCondition[] = "CONDITION";
+const char kX[] = "X";
+const char kSummarizeNum[] = "summarize_num";
+
+namespace paddle {
+namespace operators {
+
+using framework::LoDTensor;
+
+class AssertOp : public framework::OperatorBase {
+ public:
+  AssertOp(const std::string &type, const framework::VariableNameMap &inputs,
+           const framework::VariableNameMap &outputs,
+           const framework::AttributeMap &attrs)
+      : OperatorBase(type, inputs, outputs, attrs) {}
+
+ private:
+  void RunImpl(const framework::Scope &scope,
+               const platform::Place &dev_place) const override {
+    const framework::Variable *cond_var_ptr = scope.FindVar(Input(kCondition));
+    PADDLE_ENFORCE_NOT_NULL(cond_var_ptr,
+                            platform::errors::NotFound(
+                                "Input(Condition) of AssertOp is not found."));
+    const LoDTensor &cond = cond_var_ptr->Get<LoDTensor>();
+    PADDLE_ENFORCE_EQ(
+        cond.dims(), paddle::framework::make_ddim({1}),
+        platform::errors::InvalidArgument(
+            "The shape of Input(Condition) of AssertOp must be 1. But now "
+            "the Condition's shape is ",
+            cond.dims().to_str(), ".\n"));
+
+    bool cond_data = GetCondData(cond);
+    if (cond_data) {
+      return;
+    }
+
+    TensorFormatter formatter;
+    formatter.SetSummarize(Attr<int64_t>(kSummarizeNum));
+
+    const std::vector<std::string> &x_names = Inputs(kX);
+    for (const std::string &name : x_names) {
+      const framework::Variable *x_var_ptr = scope.FindVar(name);
+      const framework::LoDTensor &x_tensor = x_var_ptr->Get<LoDTensor>();
+      formatter.Print(x_tensor, name);
+    }
+
+    // Here is code to throw exception due to cond_data must be false.
+    PADDLE_ENFORCE_EQ(cond_data, true,
```

Comment:
is this check needed? looks like it always be success

Reply:
The bool `cond_data` must be false. I used PADDLE_ENFORCE_EQ to throw an exception here.

Reply:
sorry, I want to say, if check always failed, why not use `PADDLE_THROW`

Reply:
Sure, done